### PR TITLE
[SPLIT] [Rebase &&FF] [Release/202405] Miscellaneous MU_CHANGES  

### PR DESCRIPTION
**Pull request by Flickdm**
# Preface

## Description

This PR splits up https://github.com/microsoft/mu_basecore/commit/7a38833 into individual commits
to make understanding the changes easier. 

Some commits have been "stolen" for other features where they make more sense. Anything left over is being added in this PR.

---

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

Release/2023011

## Integration Instructions

N/A

### CODE REVIEW

File: MdeModulePkg/Bus/Usb/UsbBusDxe/UsbHub.c [hunk 1]
```diff
@@ -998,6 +998,23 @@ UsbRootHubInit (
...
+  } else {
+    // MU_CHANGE Add USB Hub Enumeration delay
+    EFI_TPL  OldTpl;
+
+    OldTpl = gBS->RaiseTPL (TPL_HIGH_LEVEL);
+    gBS->RestoreTPL (TPL_APPLICATION);
+
+    while (HubIf->PollCount < USB_ENUM_POLL_MINIMUM_ATTEMPTS) {
+      // wait for a minimum of USB_ENUM_POLL_MINIMUM_ATTEMPTS * (USB_ROOTHUB_POLL_INTERVAL 100 ms)
+    }
```
Comment (makubacki) [inline MdeModulePkg/Bus/Usb/UsbBusDxe/UsbHub.c]: A bit more context could optionally be provided here like:

```suggestion
    while (HubIf->PollCount < USB_ENUM_POLL_MINIMUM_ATTEMPTS) {
      // wait for a minimum of USB_ENUM_POLL_MINIMUM_ATTEMPTS * (USB_ROOTHUB_POLL_INTERVAL 100 ms)
      // with the asynchronous periodic timer interrupt that checks for hub enumeration
      // and updates PollCount on each timer trigger
    }
```

File: MdeModulePkg/Bus/Pci/PciBusDxe/PciEnumeratorSupport.c [hunk 2]
```diff
@@ -2850,7 +2850,7 @@ PciEnumeratorLight (
 **/
 EFI_STATUS
 PciGetBusRange (
-  IN     EFI_ACPI_ADDRESS_SPACE_DESCRIPTOR  **Descriptors,
+  IN OUT EFI_ACPI_ADDRESS_SPACE_DESCRIPTOR  **Descriptors, // MU_CHANGE added OUT macro
```
Comment (os-d) [inline MdeModulePkg/Bus/Pci/PciBusDxe/PciEnumeratorSupport.c:2853]: The commit title only captures part (and the unimportant part) of this change. Can you update it?

File: MdeModulePkg/Bus/Pci/XhciDxe/XhciSched.c [hunk 3]
```diff
@@ -3558,6 +3563,12 @@ XhcSetTrDequeuePointer (
   if (EFI_ERROR (Status)) {
     DEBUG ((DEBUG_ERROR, "XhcSetTrDequeuePointer: Set TR Dequeue Pointer Failed, Status = %r\n", Status));
   }
+  // MU_CHANGE - 274185
+  else {
+    Urb->Ring->RingDequeue = Urb->Ring->RingEnqueue;
+  }
+
+  // END
```
Comment (os-d) [inline MdeModulePkg/Bus/Pci/XhciDxe/XhciSched.c]: Here and above // MU_CHANGE END

File: MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.c [hunk 4]
```diff
@@ -80,6 +80,14 @@ UsbIoControlTransfer (
 
   UsbIf = USB_INTERFACE_FROM_USBIO (This);
   Dev   = UsbIf->Device;
+  // MU_CHANGE [BEGIN]
+  if (Dev->Connected == FALSE) {
+    Status = EFI_DEVICE_ERROR;
+    DEBUG ((EFI_D_ERROR, "UsbIoControlTransfer No media\n"));
```
Comment (os-d) [inline MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.c]: Across commits, use `DEBUG_ERROR` not `EFI_D_ERROR`.

File: MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h [hunk 5]
```diff
@@ -56,6 +56,14 @@ typedef struct _USB_HUB_API    USB_HUB_API;
...
 #define USB_ROOTHUB_POLL_INTERVAL  (100 * 10000U)
 #define USB_HUB_POLL_INTERVAL      64
+//
+// defines the minimum number of enumeration attempts
+//
+#define USB_ENUM_POLL_MINIMUM_ATTEMPTS  6// MU_CHANGE
+//
+// defines the maximum number of enumeration attempts
+//
+#define USB_ENUM_POLL_MAXIMUM_ATTEMPTS  200// MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h]: can you fix the space of the comment on both of these to have at least one space?

File: MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h [hunk 6]
```diff
@@ -188,15 +196,16 @@ struct _USB_DEVICE {
   USB_INTERFACE                         *ParentIf;
   UINT8                                 ParentPort; // Start at 0
   UINT8                                 Tier;
+  BOOLEAN                               Connected; // MU_CHANGE
   BOOLEAN                               DisconnectFail;
 };
 
 //
-// Stands for different functions of USB device
+// Stands for different functions of USB dev
```
Comment (os-d) [inline MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h]: This can be dropped.

File: MdeModulePkg/Core/Dxe/Image/Image.c [hunk 7]
```diff
@@ -873,12 +882,14 @@ CoreLoadPeImage (
...
+    DEBUG ((DEBUG_ERROR | DEBUG_LOAD, "%a", EfiFileName));  // &Image->ImageContext.PdbPointer[StartIndex])); // MU_CHANGE
   }
 
-  DEBUG ((DEBUG_INFO | DEBUG_LOAD, "\n"));
+  DEBUG ((DEBUG_ERROR | DEBUG_LOAD, "\n"));       // MU_CHANGE - 304324
 
-  DEBUG_CODE_END ();
+  // MU_CHANGE [BEGIN] - 304324
+  // DEBUG_CODE_END ();
+  // MU_CHANGE [END] - 304324
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Image/Image.c:933]: The commit title says add stack cookie, which is not happening here. However, I don't think this change has any value, it already was wrapped in DEBUG_CODE_BEGIN/END. So it would change to having load statement printed when debug code is not enabled. @makubacki, thoughts?

File: MdeModulePkg/Core/Dxe/Mem/Page.c [hunk 8]
```diff
@@ -2276,14 +2277,12 @@ CoreTerminateMemoryMap (
...
-            Status =  EFI_INVALID_PARAMETER;
+            DEBUG ((DEBUG_PAGE, "ExitBootServices: A RUNTIME memory entry is not on a proper alignment.\n"));    // MU_CHANGE
             goto Done;
           }
 
           if (((Entry->End + 1) & (RUNTIME_PAGE_ALLOCATION_GRANULARITY - 1)) != 0) {
-            DEBUG ((DEBUG_ERROR | DEBUG_PAGE, "ExitBootServices: A RUNTIME memory entry is not on a proper alignment.\n"));
-            Status =  EFI_INVALID_PARAMETER;
+            DEBUG ((DEBUG_PAGE, "ExitBootServices: A RUNTIME memory entry is not on a proper alignment.\n"));    // MU_CHANGE
             goto Done;
```
Comment (os-d) [inline MdeModulePkg/Core/Dxe/Mem/Page.c:2323]: I also question whether this commit is useful, I would think we want all of these prints in error cases because they are real errors we miss all the time.

**Review by apop5 — APPROVED** [review]

File: MdeModulePkg/Core/Pei/Image/Image.c [hunk 9]
```diff
@@ -761,12 +765,14 @@ PeiLoadImageLoadImage (
...
+    DEBUG ((DEBUG_ERROR | DEBUG_LOAD, "%a", EfiFileName));         // MU_CHANGE
   }
 
-  DEBUG_CODE_END ();
+  // MU_CHANGE [BEGIN]
+  // DEBUG_CODE_END ();
+  // MU_CHANGE [END]
 
-  DEBUG ((DEBUG_INFO | DEBUG_LOAD, "\n"));
+  DEBUG ((DEBUG_ERROR | DEBUG_LOAD, "\n"));
```
Comment (os-d) [inline MdeModulePkg/Core/Pei/Image/Image.c:776]: Same questions about this one @makubacki

File: MdeModulePkg/Library/UefiBootManagerLib/UefiBootManagerLib.inf [hunk 10]
```diff
@@ -126,3 +126,4 @@
   gEfiMdeModulePkgTokenSpaceGuid.PcdBootManagerMenuFile                     ## CONSUMES
   gEfiMdeModulePkgTokenSpaceGuid.PcdDriverHealthConfigureForm               ## SOMETIMES_CONSUMES
   gEfiMdeModulePkgTokenSpaceGuid.PcdMaxRepairCount                          ## CONSUMES
+  gEfiMdeModulePkgTokenSpaceGuid.PcdBootManagerInBootOrder                  ## CONSUMES ## MU_CHANGE
```
Comment (os-d) [inline MdeModulePkg/Library/UefiBootManagerLib/UefiBootManagerLib.inf]: nit add newline at EOF

File: MdeModulePkg/Library/PlatformBootManagerLibNull/PlatformBootManager.c [hunk 11]
```diff
@@ -74,3 +74,66 @@ PlatformBootManagerUnableToBoot (
...
+  OnDemandConInCOnnect
+ */
+VOID
+EFIAPI
+PlatformBootManagerOnDemandConInConnect (
+  VOID
+  )
+{
+  return;
+}
```
Comment (os-d) [inline MdeModulePkg/Library/PlatformBootManagerLibNull/PlatformBootManager.c:137]: I thought @kuqin12 had at least part of this already in PR form?

File: MdeModulePkg/Universal/CapsulePei/Capsule.h [hunk 12]
```diff
@@ -45,18 +46,18 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
```
Comment (os-d) [inline MdeModulePkg/Universal/CapsulePei/Capsule.h]: Looks like this commit can be dropped? Just adding spaces and the one header, which seems like isn't needed.

File: MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c [hunk 13]
```diff
@@ -1389,6 +1389,10 @@ ConSplitterConOutDriverBindingStart (
   //
   mConOut.TextOutMode.Mode = 0xFF;
 
+  // MU_CHANGE
+  // MU_CHANGE - Force the initial state of TextOut device to CursorInvisible.
+  TextOut->Mode->CursorVisible = FALSE;
+
```
Comment (os-d) [inline MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c:1393]: Don't need all the double MU_CHANGE tags in this commit.

File: MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c [hunk 14]
```diff
@@ -1899,7 +1903,10 @@ ConSplitterTextInAddDevice (
   //
   // Extra CheckEvent added to reduce the double CheckEvent().
   //
-  gBS->CheckEvent (TextIn->WaitForKey);
+  // MU_CHANGE
+  // MU_CHANGE - this check-event triggers the Surface on-screen keyboard to auto-activate when it shouldn't since
```
Comment (os-d) [inline MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c]: Comments should be made more generic.

File: MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c [hunk 15]
```diff
@@ -2037,7 +2044,10 @@ ConSplitterTextInExAddDevice (
   //
   // Extra CheckEvent added to reduce the double CheckEvent().
   //
-  gBS->CheckEvent (TextInEx->WaitForKeyEx);
+  // MS_CHANGE_?
+  // MSchange - this check-event triggers the Surface on-screen keyboard to auto-activate when it shouldn't since
```
Comment (os-d) [inline MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c]: MSchange and friends to MU_CHANGE

File: MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c [hunk 16]
```diff
@@ -3678,6 +3688,15 @@ ConSplitterTextInWaitForKey (
 
   Private = (TEXT_IN_SPLITTER_PRIVATE_DATA *)Context;
 
+  // MS_CHANGE_162381
+  if (!mConInIsConnect && PcdGetBool (PcdConInConnectOnDemand)) {
+    DEBUG ((EFI_D_INFO, "Connect ConIn in first WaitForKey in Lazy ConIn mode.\n"));
```
Comment (os-d) [inline MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c]: DEBUG_INFO

File: MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c [hunk 17]
```diff
@@ -5001,9 +5020,11 @@ ConSplitterTextOutClearScreen (
...
   // always be supported.
   //
-  Private->TextOutMode.CursorColumn  = 0;
-  Private->TextOutMode.CursorRow     = 0;
-  Private->TextOutMode.CursorVisible = TRUE;
+  Private->TextOutMode.CursorColumn = 0;
+  Private->TextOutMode.CursorRow    = 0;
+  // MS_CHANGE_?
+  // MSCHANGE - Default to cursor not visible
+  Private->TextOutMode.CursorVisible = FALSE;
```
Comment (os-d) [inline MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitter.c:5022]: This can be single line, with comment to right, don't need to change spacing for other two.

File: MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitterGraphics.c [hunk 18]
```diff
@@ -613,10 +613,11 @@ TextOutSetMode (
...
-  Private->TextOutMode.CursorColumn  = 0;
-  Private->TextOutMode.CursorRow     = 0;
-  Private->TextOutMode.CursorVisible = TRUE;
-
+  Private->TextOutMode.Mode         = (INT32)ModeNumber;
+  Private->TextOutMode.CursorColumn = 0;
+  Private->TextOutMode.CursorRow    = 0;
+  // MU_CHANGE
+  // MU_CHANGE - Initialize the console with the cursor not visible.
+  Private->TextOutMode.CursorVisible = FALSE;
```
Comment (os-d) [inline MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitterGraphics.c:621]: Same.

File: MdeModulePkg/Universal/HiiDatabaseDxe/Font.c [hunk 19]
```diff
@@ -2132,7 +2132,7 @@ HiiStringToImage (
         UINT32  Bkgnd = Background.Blue | Background.Green << 8 | Background.Red << 16;
         BltBuffer = AllocatePool (RowInfo[RowIndex].LineWidth * RowInfo[RowIndex].LineHeight * sizeof (EFI_GRAPHICS_OUTPUT_BLT_PIXEL));
         SetMem32 (BltBuffer, RowInfo[RowIndex].LineWidth * RowInfo[RowIndex].LineHeight * sizeof (EFI_GRAPHICS_OUTPUT_BLT_PIXEL), Bkgnd);
-        // MU_CHANGE [END] 
+        // MU_CHANGE [END]
```
Comment (os-d) [inline MdeModulePkg/Universal/HiiDatabaseDxe/Font.c:2135]: Should be fixed in original commit.

File: MdeModulePkg/Universal/Variable/RuntimeDxe/Variable.c [hunk 20]
```diff
@@ -560,6 +567,9 @@ Reclaim (
   CommonUserVariableTotalSize = 0;
   HwErrVariableTotalSize      = 0;
 
+  // MU_CHANGE - This may be specific to the MS implementation.
+  DEBUG ((DEBUG_INFO, "%a Reclaim variables started.\n", __FUNCTION__));
```
Comment (os-d) [inline MdeModulePkg/Universal/Variable/RuntimeDxe/Variable.c]: `__func__`

File: MdePkg/Include/Library/DebugLib.h [hunk 21]
```diff
@@ -64,6 +65,7 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #define EFI_D_INFO      DEBUG_INFO
 #define EFI_D_DISPATCH  DEBUG_DISPATCH
 #define EFI_D_VARIABLE  DEBUG_VARIABLE
+#define EFI_D_SMI       DEBUG_SMI  // MU_CHANGE - 141550 Added for SMI auditing options.
 #define EFI_D_BM        DEBUG_BM
```
Comment (os-d) [inline MdePkg/Include/Library/DebugLib.h]: Does this get used anywhere? If not we should drop

File: MdePkg/Include/IndustryStandard/PeImage.h [hunk 22]
```diff
@@ -117,6 +117,34 @@ typedef struct {
...
+  UINTN    ProcessHeapFlags;
+  UINTN    ProcessAffinityMask;
+  UINT8    CSDVersion;
+  UINT8    Reserved1;
+  UINTN    EditList;                        // VA
+  UINTN    SecurityCookie;
+  VOID     **SEHandlerTable;
+  UINTN    SEHandlerCount;
+} EFI_IMAGE_LOAD_CONFIG_DIRECTORY;
+// MU_CHANGE [END]
```
Comment (os-d) [inline MdePkg/Include/IndustryStandard/PeImage.h]: Does this get used anywhere? If not, we should drop.

File: MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h [hunk 23]
```diff
@@ -56,6 +56,14 @@ typedef struct _USB_HUB_API    USB_HUB_API;
 //
 #define USB_ROOTHUB_POLL_INTERVAL  (100 * 10000U)
 #define USB_HUB_POLL_INTERVAL      64
+//
+// defines the minimum number of enumeration attempts
+//
+#define USB_ENUM_POLL_MINIMUM_ATTEMPTS  6// MU_CHANGE
```
Comment (kenlautner) [inline MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h]: nit:  Probably should have a space between the value and // MU_CHANGE.

File: MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h (hunk 5 above)
Comment (kenlautner) [inline MdeModulePkg/Bus/Usb/UsbBusDxe/UsbBus.h]: Same as above.

**Review by kenlautner — APPROVED** [review]

File: MdeModulePkg/Library/PlatformBootManagerLibNull/PlatformBootManager.c (hunk 11 above)
Comment (Flickdm) [reply MdeModulePkg/Library/PlatformBootManagerLibNull/PlatformBootManager.c:137]: If that's true I'm happy to give @kuqin12 this commit just so it can go with like changes.

File: MdeModulePkg/Bus/Pci/XhciDxe/XhciSched.c [hunk 24]
```diff
@@ -3558,6 +3563,12 @@ XhcSetTrDequeuePointer (
   if (EFI_ERROR (Status)) {
     DEBUG ((DEBUG_ERROR, "XhcSetTrDequeuePointer: Set TR Dequeue Pointer Failed, Status = %r\n", Status));
   }
+  // MU_CHANGE [BEGIN] - 274185
+  else {
+    Urb->Ring->RingDequeue = Urb->Ring->RingEnqueue;
```
Comment (kuqin12) [inline MdeModulePkg/Bus/Pci/XhciDxe/XhciSched.c:3568]: Is there any information on why this change is needed?

File: MdeModulePkg/Bus/Usb/UsbMouseAbsolutePointerDxe/UsbMouseAbsolutePointer.c [hunk 25]
```diff
@@ -847,7 +847,7 @@ OnMouseInterruptComplete (
     UsbMouseAbsolutePointerDevice->State.CurrentZ =
       MIN (
         MAX (
-          (INT64)UsbMouseAbsolutePointerDevice->State.CurrentZ + *((INT8 *)Data + 1),
+          (INT64)UsbMouseAbsolutePointerDevice->State.CurrentZ + *((INT8 *)Data + 3),      // MU_CHANGE changed to +3 from +1
```
Comment (kuqin12) [inline MdeModulePkg/Bus/Usb/UsbMouseAbsolutePointerDxe/UsbMouseAbsolutePointer.c:850]: Do you have any info on why 3?

File: MdeModulePkg/Library/PlatformBootManagerLibNull/PlatformBootManager.c (hunk 11 above)
Comment (kuqin12) [reply MdeModulePkg/Library/PlatformBootManagerLibNull/PlatformBootManager.c:137]: That was for the content in silicon ARM. This change is still needed.

File: MdeModulePkg/Universal/Variable/RuntimeDxe/Variable.c [hunk 26]
```diff
@@ -53,6 +53,13 @@ VARIABLE_INFO_ENTRY  *gVariableInfo = NULL;
 ///
 BOOLEAN  mEndOfDxe = FALSE;
 
+// MU_CHANGE [BEGIN] - 279849
+///
+/// The flag to indicate if a Reclaim has been done at Runtime.
+///
+BOOLEAN  mReclaimedAtRuntime = FALSE;
```
Comment (kuqin12) [inline MdeModulePkg/Universal/Variable/RuntimeDxe/Variable.c:60]: FYI, this is to mitigate the number of runtime reclaim for wear leveling sake. In case you want to update the commit message.

**Review by kuqin12 — APPROVED** [review]

File: MdeModulePkg/Universal/Variable/RuntimeDxe/Variable.c (hunk 26 above)
Comment (Flickdm) [reply MdeModulePkg/Universal/Variable/RuntimeDxe/Variable.c:60]: I'll go ahead and update the commit message to include that information / history

File: MdeModulePkg/Bus/Usb/UsbMouseAbsolutePointerDxe/UsbMouseAbsolutePointer.c (hunk 25 above)
Comment (Flickdm) [reply MdeModulePkg/Bus/Usb/UsbMouseAbsolutePointerDxe/UsbMouseAbsolutePointer.c:850]: No idea :)

File: MdeModulePkg/Core/DxeIplPeim/DxeLoad.c [hunk 27]
```diff
@@ -147,7 +151,10 @@ InstallIplPermanentMemoryPpis (
   IN VOID                       *Ppi
   )
 {
-  EFI_STATUS              Status;
+  EFI_STATUS  Status;
+
+  // MS_CHANGE_190466
```
Comment (VivianNK) [inline MdeModulePkg/Core/DxeIplPeim/DxeLoad.c]: `MS_CHANGE` should be replaced with `MU_CHANGE` in 2 places in this file
